### DISCUSSION TITLE
link prefetch in next 13 /pages - constantly pre-fetching empty data

### DISCUSSION THREAD

**Discussion by spence-s** [vercel/next.js#44596] [Help] [answered]
In next 12 links would simply pre-fetch the .js files for the page that was being navigated to. 

Now, after upgrading, our app issues a request for the .js file and another prefetch for a .json file/_data route. We are only using `getServerSideProps` on these pages. This essentially caused us to ddos ourselves because our e-commerce site has hundreds of links on every page, and a single user is issuing _thousands_ of requests for this empty json data. 

In every single case, the .json data is just an empty object. `{}` and provides no value. 

We added `prefetch={false}` to all of our links, which improves the amount of requests, but next still issues all the empty data requests on every hover. 

How on earth can we turn these empty data requests off?? we were happy with the previous prefetch behavior.

For now we are going to roll our own link solution on top of next/router, but we'd rather use what the framework offers. Why this terrible change of behavior for us. Is this a bug??

**Comment by spence-s** [comment, marked as answer]
Well - we realized this behavior was because we also added a middleware and so these .json fetches are actually pre-fetching the middleware data or something along those lines. 

set config.experimental.middlewarePrefetch = 'strict' to avoid this behavior

**Comment by pawelgalazka** [comment]
Worth pointing out that default prefetch behaviour described here can break pages which are behind the cache. What I noticed for my case is that prefetch responses for page `json` files are also stripped from `cache-control` header even if `getServerSideProps` is used for getting page data. So in combination with returning empty responses, caching service can cache prefetch requests with empty responses and when actual request is made during page transition to get page data `json`, empty response is returned as well. This breaks the app.

**Reply by fasani-tx** [reply]
I have this same issue on my staging environment. I opened a discussion here:
https://github.com/vercel/next.js/discussions/45044

I also have middleware and had to move the middleware up one folder from `pages` to `src`.

**Reply by fasani-tx** [reply]
I noticed a new (??) header in Next 13 (`x-middleware-prefetch: 1`).

I have not tested this, but I guess our cache setup strips this header.

2 commits reference this header:
https://github.com/vercel/next.js/commit/533c242094a08a41bd8cc89958c438998fbc27f0
https://github.com/vercel/next.js/commit/2f744a1ae32a65f1362af0957e3c57f33d102914
